### PR TITLE
Add a warning message when detecting cyclic dependencies in mutations

### LINKED ISSUE #53457: Cyclic mutation dependency crashes game with no error or log
### Describe the bug

Cyclic dependencies in mutation definitions apparently are not checked for. In my example I had forgotten to update the "changes_to" field in a sequence of mutations resulting in a mutation which "changes into itself". This did not throw any errors upon loading the character (this definition was in a mod, in case that matters). Attempting to gain the mutation via debug menu does not throw an error either, instead the game freezes for a moment then crashes to desktop without producing any log output.

Worse in this particular case it seems the mutation is queried when listed in the "changes_to" of a *different* mutation which can quickly obscure the true cause due to the lack of feedback. Imagine someone made this mistake with "MUT_TOUGH2" crashing the game when gaining the prerequisite "MUT_TOUGH" which lists over a dozen other mutations across various fields.

I have not tested the plethora of other cases, self-"prereqs" and self-"leads_to" seem obvious as well while "threshreq" and "category" might be difficult to loop without dedicated sabotage.

### Steps to reproduce

This is a close-to-minimal "working" example. Note the identical ID in "id" and "changes_to":
```json
  {
    "type": "mutation",
    "id": "CRASHME",
    "name": { "str": "Foo" },
    "points": 1,
    "description": "Crashes the game when gained via debug p->M.",
    "changes_to": [ "CRASHME" ],
    "mixed_effect": true
  }
```

### Expected behavior

Should probably throw an error on mod load as this kind of inconsistency can be checked for immediately from the JSON data. Possibly upon loading *all* mods in case several individually inoffensive mods create a dependency loop. At the very least the game should crash *with* a legible error message upon attempting to gain the mutation.

### Screenshots

_No response_

### Versions and configuration

- OS: Windows
    - OS Version: 10.0 2009
- Game Version: cf9e61e [64-bit]
- Graphics Version: Tiles
- Game Language: System language []
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    No Fungal Growth [no_fungal_growth],
    Bionic Professions [package_bionic_professions],
    Stats Through Skills [StatsThroughSkills],
    Weird Science [Ps_WeirdScience]
]
"Ps_WeirdScience" is what I used to insert the faulty mutation.

### Additional context

Please independently verify the lack of meaningful errors. My crash.log hasn't been updated for 7 days and the debug.log just gets spammed with "invalid overmap special id" errors because the debug world is a little older.

### LINKED ISSUE #55151: Game quits with segmentation fault if mutation has circular dependency
### Describe the bug

If the "Saprophage" mutation JSON is altered so its "prereqs" includes another mutation that requires "Saprophage" itself (creating a circular dependency), the game crashes with a segmentation fault when acquiring the mutation.

### Steps to reproduce

1. Change the "Saprophage" mutation JSON, adding "ROOTS3" to its "prereqs":

```diff
     "id": "SAPROPHAGE",
-    "prereqs": [ "ROOTS2" ],
+    "prereqs": [ "ROOTS2", "ROOTS3" ],
```

2. Load game, and acquire either the "Saprophage" or "Rooter" mutations. Segmentation fault occurs.

### Expected behavior

Game should not crash.

Mutations with circular or otherwise unresolvable dependencies should be detected at JSON load/validation time.

### Screenshots

_No response_

### Versions and configuration

- OS: Linux
    - OS Version: Distributor ID: Pop; Description: Pop!_OS 20.04 LTS; Release: 20.04; Codename: focal; 
- Game Version: 0.F-5861-g3111ca2fdf [64-bit]
- Graphics Version: Tiles
- Game Language: System language []
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    No Fungal Growth [no_fungal_growth],
    Bionic Professions [package_bionic_professions]
]

### Additional context

No crash.log is generated, and [debug.log](https://github.com/CleverRaven/Cataclysm-DDA/files/8010443/debug.log) has nothing useful in it.

Bug reported by OmniscientQ on discord

### PR DESCRIPTION
**Pull request by SandNoodle**
#### Summary
Bugfixes "Resolve circular dependencies for invalid mutations by adding a warning message on save load when one occurs"

#### Purpose of change

Fixes #53457 
Fixes #55151

#### Describe the solution

The solution is twofold:
1. Adds warning message on save load when self referencing mutation is included (as per #53457)
2. (Optionally) Try to remove circular dependencies or at least prevent stack overflow (as per #55151)

#### Describe alternatives you've considered

Currently none.

#### Testing

1. Include the code snippet from #53457
2. Try to load a save
3. Get a warning (self-reference)

and if (optional) change 2 is implemented:

1. Include code snippet from #55151
2. Try to load a save
3. Get a warning (deeply nested circular dependency)
4. Go to the debug menu (p -> M)
5. Gain Saprophage mutation

### CODE REVIEW

**Comment by SandNoodle** [comment]
I'm going to convert this to draft, there's still some work to do.

TODO:
- [ ] Currently, the warning message is displayed when mutation references only itself - it does not, however, display a warning message when circular dependency is deeply nested; i.e. A -> B -> C -> A (as in Saprophage's case).
- [ ] Saprophage does not crash the game by stack overflow, but the "CRASHME" snippet still does.

**Comment by BrettDong** [comment]
How about using a depth-first search to detect cycles in dependencies?

**Comment by SandNoodle** [comment]
> How about using a depth-first search to detect cycles in dependencies?

That's a good idea, but sadly I'm not familiar with this alghortim (and the past two days implementing it were fruitless in that regard), so I'd need a bit of help with implementing it.

**Comment by Willenbrink** [comment]
I'm willing to help if you tell me exactly what you need help with.

First things first, I don't think the game should crash in this case.
The general idea to check for circular dependencies is:
* Each mutation has a list of dependencies (here that's `mvec` if I'm not mistaken). As they each have their own dependencies this forms a tree. (A mutation might occur as a dependency in multiple places so its actually a directed acyclic graph but you can treat it as a tree without problems)
* A circular dependency occurs whenever a mutation is contained somewhere in its own dependency tree (as it's then no longer a real tree)
* So the simplest approach would be: for every mutation check whether it occurs somewhere in its dependencies via something like `contains(mid, mvec)`. Don't actually do this as that will be quite slow (Not sure how many mutations exactly exist and how large the dependency trees are. Its probably not really a problem but it might become one in the future).
* Instead you can keep a list of known mutations and traverse this tree in a depth-first manner. Depth first simply means that you enter the first grandchild before the second child. It would work something like
```cpp
bool check(already_visited, mid) {
  if(already_visited.contains(mid))
    return true; // Circular dependency
  already_visited.add(mid);
  for(mutation : mvec)
    // We check mutation completely before moving on to the next element of mvec.
    // Depth-first as opposed to breadth-first!
    check(already_visited, mutation);
  already_visited.remove(mid);
  return false;
}
```
* In my opinion recursion works nicely for this. If you create a copy of `already_visited` for every call instead of passing a reference you don't have to worry about mutation at all (so the remove is unnecessary). You can also use a worklist/stack based approach although I find them to be a bit harder to understand.

**Comment by SandNoodle** [comment]
So basically, I have a problem with these two things:
- Implementing the DSF (The logic was surprisingly easy to work out, programming, not so much).
- Finding where all the mutations are stored inside the code so I can check them all at once and remove any cyclic dependencies.

Due to that I went for the low hanging fruit and implemented the warning message.

**Comment by Willenbrink** [comment]
I think the check can be simply added to the already existing consistency checks (as you have done) and maybe be done in a separate function.
Then you wouldn't check them all at once but simply extend the current checks. I guess they are already checked all at once?

Adapting my DFS idea from above, that would look something like:
```cpp
bool check(std::vec<trait_id> already_visited, trait_id current) {
  if(already_visited.contains(current.id))
    return true; // Circular dependency
  already_visited.push_back(current.id);
  for(trait_id prereq : current.prereqs) //also for prereqs2
    // We check each prerequisite completely before moving on to the next element of .
    // Depth-first as opposed to breadth-first!
    check(already_visited, prereq);
  return false;
}
```

**Comment by SandNoodle** [comment]
Thanks for help, @Willenbrink and @BrettDong. 

I'll mark this PR as ready to merge, as the basic functionality was implemented (warning message on save loading), so if someone wants to pick this up further I can make a new issue.

File: src/mutation_data.cpp [hunk 1]
```diff
@@ -567,13 +567,51 @@ std::string mutation_branch::desc() const
...
+    }
+
+    return false;
+}
+
+static void check_has_cyclic_dependency( const trait_id &mid )
+{
+    std::vector<trait_id> already_visited;
+    if( ::has_cyclic_dependency( mid, already_visited ) ) {
+        debugmsg( "mutation %s references itself in either of it's prerequsition branches.  Program will crash if player would gain this mutation.",
```
Comment (Willenbrink) [inline src/mutation_data.cpp:598]: Minor rephrasing: "Mutation %s references itself in either of its prerequisite branches.  The program will crash if the player gains this mutation."

File: src/mutation_data.cpp [hunk 2]
```diff
@@ -567,13 +567,51 @@ std::string mutation_branch::desc() const
...
+        }
+    }
+
+    return false;
+}
+
+static void check_has_cyclic_dependency( const trait_id &mid )
+{
+    std::vector<trait_id> already_visited;
+    if( ::has_cyclic_dependency( mid, already_visited ) ) {
```
Comment (Willenbrink) [inline src/mutation_data.cpp:597]: The double colon is used to refer to global variables. So it is necessary for `::check_consistency` as there are two, `mutation_branch::check_consistency` and `static void check_consistency` and we want the latter here (the global one). It's not necessary for `check_has_cyclic dependency`.

File: src/mutation_data.cpp [hunk 3]
```diff
@@ -654,7 +692,9 @@ void mutation_branch::check_consistency()
             }
         }
 
-        ::check_consistency( mdata.prereqs, mid, "prereq" );
+        ::check_has_cyclic_dependency( mid );
```
Comment (Willenbrink) [inline src/mutation_data.cpp:695]: No `::` necessary.

File: src/mutation_data.cpp [hunk 4]
```diff
@@ -567,13 +567,51 @@ std::string mutation_branch::desc() const
...
 static void check_consistency( const std::vector<trait_id> &mvec, const trait_id &mid,
                                const std::string &what )
 {
     for( const auto &m : mvec ) {
         if( !m.is_valid() ) {
             debugmsg( "mutation %s refers to undefined %s %s", mid.c_str(), what.c_str(), m.c_str() );
         }
+
+        if( m == mid ) {
+            debugmsg( "mutation %s refers to itself in %s context.  Program will crash if player would gain this mutation",
```
Comment (Willenbrink) [inline src/mutation_data.cpp:612]: Minor rephrasing: "Mutation %s references itself in %s context. The program will crash if the player gains this mutation."

**Review by Willenbrink — COMMENTED** [review]
Looks good to me. Would be nice if you could add the suggested changes.

**Review by Maleclypse — COMMENTED** [review]
Json looks good at least

**Pull request merged**